feat: extend api/v2/event-types

ibex088 opened the pull request:
### Overview

This PR introduces new attributes to the following updated endpoints:

### Event Types
- **[POST]** `/v2/event-types`
- **[PATCH]** `/v2/event-types/{eventTypeId}`
- **[GET]** `/v2/event-types/{eventTypeId}` — now includes the newly added properties.

### Organizations Event Types
- **[POST]** `/v2/organizations/{orgId}/teams/{teamId}/event-types`
- **[PATCH]** `/v2/organizations/{orgId}/teams/{teamId}/event-types/{eventTypeId}`
- **[GET]** `/v2/organizations/{orgId}/teams/{teamId}/event-types/{eventTypeId}` — now includes the newly added properties.


---

### How to Test
1. set   `"version": "1.2.3"` in packages/platform/libraries/package.json
2. add     `"@calcom/platform-libraries-1.2.3": "npm:@calcom/platform-libraries@1.2.3"` to apps/api/v2/package.json

#### Testing `api/v2/event-types`:

1. **Limit Booking Frequency:**
![Screenshot 2024-08-12 at 11 02 19 AM](https://github.com/user-attachments/assets/9c83036b-9ec3-4344-91f5-c4462c92029d)

   - Input for the handler:
     ```json
     "bookingLimits": {
       "PER_DAY": 1,
       "PER_WEEK": 11,
       "PER_MONTH": 22,
       "PER_YEAR": 33
     }
     ```
   - API Endpoint:
     ```json
     "bookingLimitsCount": {
       "day": 1,
       "week": 11,
       "month": 22,
       "year": 33
     }
     ```

3. **Limit Total Booking Duration:**
![Screenshot 2024-08-12 at 11 03 06 AM](https://github.com/user-attachments/assets/b64d24f6-a333-4de2-8be1-c4b182fd1702)

   - Input for the handler:
     ```json
     "durationLimits": {
       "PER_DAY": 60,
       "PER_WEEK": 120,
       "PER_MONTH": 240,
       "PER_YEAR": 360
     }
     ```
   - API Endpoint:
     ```json
     "bookingLimitsDuration": {
       "day": 60,
       "week": 120,
       "month": 240,
       "year": 360
     }
     ```

4. **Limit Future Bookings (Booking Window):**
   - **Range Type:**
![Screenshot 2024-08-12 at 11 03 49 AM](https://github.com/user-attachments/assets/1bbcac9a-debd-4bed-a1ed-32344d1b0b05)

     - Input for the handler:
       ```json
       {
         "periodType": "RANGE",
         "periodStartDate": "2024-08-06T09:14:30.000Z",
         "periodEndDate": "2024-08-23T18:30:00.000Z"
       }
       ```
     - API Endpoint:
       ```json
       "bookingWindow": {
         "type": "range",
         "value": ["2024-08-06", "2024-08-28"]
       }
       ```

   - **Rolling Type (Calendar Days):**
![Screenshot 2024-08-12 at 11 04 18 AM](https://github.com/user-attachments/assets/ec2b3d76-34bd-4e6e-af80-343cb4f57e34)

     - Input for the handler:
       ```json
       {
         "periodType": "ROLLING",
         "periodDays": 30,
         "periodCountCalendarDays": true
       }
       ```
     - API Endpoint:
       ```json
       "bookingWindow": {
         "type": "calendarDays",
         "value": 30,
         "rolling": false
       }
       ```

   - **Rolling Window Type (Calendar Days):**
![Screenshot 2024-08-12 at 11 04 44 AM](https://github.com/user-attachments/assets/2a49cde3-f45f-46d6-b3bd-90398915f036)

     - Input for the handler:
       ```json
       {
         "periodType": "ROLLING_WINDOW",
         "periodDays": 30,
         "periodCountCalendarDays": true
       }
       ```
     - API Endpoint:
       ```json
       "bookingWindow": {
         "type": "calendarDays",
         "value": 30,
         "rolling": true
       }
       ```

   - **Rolling Type (Business Days):**
![Screenshot 2024-08-12 at 11 05 05 AM](https://github.com/user-attachments/assets/20d46608-2f82-4b31-80e4-7b904f877203)

     - Input for the handler:
       ```json
       {
         "periodType": "ROLLING",
         "periodDays": 30,
         "periodCountCalendarDays": false
       }
       ```
     - API Endpoint:
       ```json
       "bookingWindow": {
         "type": "businessDays",
         "value": 30,
         "rolling": false
       }
       ```

   - **Rolling Window Type (Business Days):**
![Screenshot 2024-08-12 at 11 05 16 AM](https://github.com/user-attachments/assets/d8672fef-5dde-4dce-a120-0cd49f1e6310)

     - Input for the handler:
       ```json
       {
         "periodType": "ROLLING_WINDOW",
         "periodDays": 30,
         "periodCountCalendarDays": false
       }
       ```
     - API Endpoint:
       ```json
       "bookingWindow": {
         "type": "businessDays",
         "value": 30,
         "rolling": true
       }
       ```
5. Offset start times
![Screenshot 2024-08-12 at 11 10 50 AM](https://github.com/user-attachments/assets/e083f62c-2d7c-4347-9281-4bb6667f5edb)

     - Input for the handler:
       ```json
       {
         "offsetStart": 20,
       }
       ```
     - API Endpoint:
       ```json
       "offsetStart": 20
       ```
6. Only show the first slot of each day as available
![Screenshot 2024-08-12 at 11 12 06 AM](https://github.com/user-attachments/assets/9a47fefa-5469-471e-a7cc-b1af6d771eb8)

     - Input for the handler:
       ```json
       {
         "onlyShowFirstAvailableSlot": true,
       }
       ```
     - API Endpoint:
       ```json
       "onlyShowFirstAvailableSlot": true
       ```
---

### Mandatory Tasks (DO NOT REMOVE)
- [x] I have self-reviewed the code (A decent size PR without self-review might be rejected).
- [x] N/A - I have added a Docs issue [here](https://github.com/calcom/docs/issues/new) if this PR makes changes that would require a [documentation change](https://docs.cal.com). If N/A, write N/A here and check the checkbox.
- [x] I confirm automated tests are in place that prove my fix is effective or that my feature works.

ibex088 committed:
feat: extended api/v2/event-types to add more attributes

- additional attributes added
* bookingLimits
* onlyShowFirstAvailableSlot
* durationLimits
* offsetStart
* bookingWindow

ibex088 committed:
add swagger docs

ibex088 committed:
type fix

ibex088 committed:
extend api/v2/teams/[teamId]/event-types

ibex088 committed:
added tests

ibex088 committed:
Merge branch 'main' into feat/extend-v2-event-types

ibex088 committed:
extend api/v2/orgs/[orgId]/teams/[teamid]/event-types

ThyMinimalDev committed:
Merge branch 'main' into feat/extend-v2-event-types

ThyMinimalDev commented on apps/api/v2/src/ee/event-types/event-types_2024_06_14/controllers/event-types.controller.e2e-spec.ts:
```diff
@@ -232,7 +248,11 @@ describe("Event types Endpoints", () => {
           expect(createdEventType.ownerId).toEqual(user.id);
           expect(createdEventType.scheduleId).toEqual(firstSchedule.id);
-
+          expect(createdEventType.bookingLimits).toEqual(body.bookingLimits);
+          expect(createdEventType.onlyShowFirstAvailableSlot).toEqual(body.onlyShowFirstAvailableSlot);
+          expect(createdEventType.durationLimits).toEqual(body.durationLimits);
+          expect(createdEventType.offsetStart).toEqual(body.offsetStart);
+          expect(createdEventType.bookingWindow).toEqual(body.bookingWindow);
```
can we have similar test for update and get endpoint

SomayChauhan replied:
added 👍

ibex088 committed:
update: added tests to get and update event-type as well

supalarry commented on apps/api/v2/src/ee/event-types/event-types_2024_06_14/services/input-event-types.service.ts:
```diff
@@ -4,6 +4,10 @@ import {
+import {
+  transformApiEventTypeIntervalLimits,
+  transformApiEventTypeFutureBookingLimits,
+} from "@calcom/platform-libraries-1.2.3";
```
! Before merging to main need to re-publish libraries and update import

supalarry commented on apps/api/v2/src/ee/event-types/event-types_2024_06_14/services/output-event-types.service.ts:
```diff
@@ -12,6 +12,12 @@ import {
+  parseBookingLimit,
+  getResponseEventTypeIntervalLimits,
+  getResponseEventTypeFutureBookingLimits,
+} from "@calcom/platform-libraries-1.2.3";
```
! Before merging to main need to re-publish libraries and update import

supalarry commented on packages/lib/event-types/transformers/api-request.ts:
```diff
@@ -102,6 +117,51 @@ function transformApiEventTypeBookingFields(
   return customBookingFields;
 }
 
+function transformApiEventTypeIntervalLimits(
```
Would be good to add tests of functions added to `api-request.ts` in `api-request.spec.ts` - it should be fast to add and would be good because transformation logic can sometimes be hard to grasp right away.

SomayChauhan replied:
added https://github.com/calcom/cal.com/pull/16128/commits/b8bf222e894147dc9d0fc055443188cecc25cf29

supalarry replied:
Great!

supalarry commented on packages/lib/event-types/transformers/api-response.ts:
```diff
@@ -189,4 +201,63 @@ function getResponseEventTypeBookingFields(
 }
 
-export { getResponseEventTypeLocations, getResponseEventTypeBookingFields };
+function getResponseEventTypeIntervalLimits(
```
Would be good to add tests of functions added to `api-response.ts` in `api-response.spec.ts` - it should be fast to add and would be good because transformation logic can sometimes be hard to grasp right away.

SomayChauhan replied:
added https://github.com/calcom/cal.com/pull/16128/commits/b8bf222e894147dc9d0fc055443188cecc25cf29

supalarry replied:
Amazing!

supalarry reviewed (commented):
Amazing work Somay! I have some thoughts below let me know what you think.

SomayChauhan commented on packages/platform/types/event-types/event-types_2024_06_14/inputs/enums/booking-window.enum.ts:
```diff
@@ -0,0 +1,11 @@
+}
+
+export enum BookingWindowPeriodOutputTypeEnum_2024_06_14 {
+  RANGE = "RANGE",
```
these are for saving on our DB, we don't take them as input or show them as output anywhere
we take BookingWindowPeriodInputTypeEnum_2024_06_14 as input.

ibex088 committed:
rename variables

- bookingLimits -> bookingLimitsCount
- durationLimits -> bookingLimitsDuration

ibex088 committed:
refactor: rename bookingLimits and durationLimits, add descriptions

- Renamed `bookingLimits` to `bookingLimitsCount`
- Renamed `durationLimits` to `bookingLimitsDuration`
- Added descriptions with examples to `bookingLimitsCount` and `bookingLimitsDuration`

ibex088 committed:
Merge branch 'main' into feat/extend-v2-event-types

ibex088 committed:
feat: add unit tests for transformation logic in api-request and api-response

- Added tests for `transformApiEventTypeIntervalLimits`
- Added tests for `transformApiEventTypeFutureBookingLimits`

ibex088 committed:
fix: type errors

SomayChauhan commented on packages/platform/types/event-types/event-types_2024_06_14/inputs/update-event-type.input.ts:
```diff
@@ -58,6 +60,27 @@ export class UpdateEventTypeInput_2024_06_14 {
+
+  @IsOptional()
+  @Type(() => IntervalLimits_2024_06_14)
+  bookingLimits?: IntervalLimits_2024_06_14;
```
added om https://github.com/calcom/cal.com/pull/16128/commits/d6f669548fcac33887fcf04faf76578e9f617c47

supalarry replied:
Looks good!

SomayChauhan commented on packages/platform/types/event-types/event-types_2024_06_14/inputs/update-event-type.input.ts:
```diff
@@ -58,6 +60,27 @@ export class UpdateEventTypeInput_2024_06_14 {
+
+  @IsOptional()
+  @Type(() => IntervalLimits_2024_06_14)
+  durationLimits?: IntervalLimits_2024_06_14;
```
renamed in https://github.com/calcom/cal.com/pull/16128/commits/d6f669548fcac33887fcf04faf76578e9f617c47

supalarry replied:
Looks good!

ibex088 committed:
feat: added custom validators for BookingLimitsCount and BookingLimitsDuration

ibex088 committed:
Merge branch 'main' into feat/extend-v2-event-types

ibex088 committed:
fix: unit-tests

ibex088 committed:
Merge branch 'main' into feat/extend-v2-event-types

ibex088 committed:
Update CHANGELOG.md

ibex088 committed:
refactor: update imports to use platform-libraries from @calcom/platform-libraries-1.2.3

supalarry reviewed (dismissed):
Excellent PR! Great tests, clean code and you did great when navigating platform related code!

SomayChauhan commented:
Thanks a lot for all the support throughout! 💯

ibex088 committed:
fix: prevent double transformation of bookingWindow and bookingFields

- resolving e2e test failures

ThyMinimalDev committed:
Merge branch 'main' into feat/extend-v2-event-types

ThyMinimalDev commented on apps/api/v2/src/ee/event-types/event-types_2024_06_14/services/output-event-types.service.ts:
```diff
@@ -80,6 +95,15 @@ export class OutputEventTypesService_2024_06_14 {
+      periodCountCalendarDays: databaseEventType.periodCountCalendarDays,
+      periodStartDate: databaseEventType.periodStartDate,
+      periodEndDate: databaseEventType.periodEndDate,
+    } as TransformFutureBookingsLimitSchema_2024_06_14);
```
instead of `as` we should use `satisfies`

shirazdole commented:
Banger and really easy ticket to follow w/ the screenshots! Thank you @SomayChauhan
